2단계 - 연관 관계 매핑

victor-jo opened the pull request:
2단계 - 연관 관계 매핑

잘 부탁드립니다!

victor-jo committed:
refactor(answer): 작성자에 대해 연관관계로 처리

- OneToOne 연관관계로 추가
- 테스트케이스에서는 항상 새로운 유저로 저장 (영속화 되지 않은 엔티티인데 아이디가 있으면 안되는 오류)

victor-jo committed:
refactor(test): JPA 테스트시 Auto-Increment 로 테스트가 꼬이는 현상 해결

- H2 DB 안의 Auto-Increment 값이 계속 남아있어 시퀀스가 증가되는 현상 수정

victor-jo committed:
refactor(answer): 등록자 연관관계 변경

- 등록자 한명은 여러개의 답변을 달 수도 있다

victor-jo committed:
refactor(answer): 질문 엔티티 연관관계 설정

- 질문 엔티티 하나에는 여러개의 답변이 달릴 수 있다
- 답변은 질문 엔티티에서 참조 객체로 불러오도록 한다
- 답변 테스트에서 질문을 계속 생성하지 않고 한번 생성된 질문을 재사용

victor-jo committed:
refactor(question): 질문 엔티티에서 삭제된 답변은 가져오지 않도록 리팩터링

- 양방향 참조에서 서로 객체 제거

victor-jo committed:
refactor(question): 질문 엔티티 작성자 연관관계 추가

victor-jo committed:
refactor(deleteHistory): 삭제 히스토리 작성자에 연관관계 매핑 추가

minzzang commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -56,12 +58,12 @@ public boolean equals(Object o) {
                 contentType == that.contentType &&
                 Objects.equals(contentId, that.contentId) &&
-                Objects.equals(deletedById, that.deletedById);
+                Objects.equals(whoDeleted, that.whoDeleted);
```
혹시 equlas 비교하는 변수 설정하는 기준이 있을까요~?

victor-jo replied:
찾아보니 QnA Service Test 코드에서

new DeleteHistory(ContentType.QUESTION, question.getId(), question.getWriter(), LocalDateTime.now()),
new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter(), LocalDateTime.now())

두개 객체와 verify 메소드로 확인하네요!👍

해당 부분 Equals 에서 CreatedAt 은 매번 변하니 검사하면 안되고

ContentType, ContentId, Writer 비교를 수행하면 될 것 같아 그렇게 처리하였습니다!

minzzang replied:
저장하면 엔티티마다 고유한 id를 갖기 때문에 id 값으로 비교를 해도 괜찮을 것 같습니다~!
https://velog.io/@park2348190/JPA-Entity%EC%9D%98-equals%EC%99%80-hashCode

minzzang commented on src/main/java/qna/domain/Question.java:
```diff
@@ -1,23 +1,34 @@
     private boolean deleted = false;
+    @OneToMany(mappedBy = "question")
+    @Where(clause = "deleted = false")
+    private List<Answer> answers = new ArrayList<>();
```
일급 컬렉션으로 변경해보면 어떨까요~?

victor-jo replied:
넵 일급 컬렉션으로 변경하여 추가적인 메시지 처리 가능하도록 수정하였습니다!

minzzang commented on src/test/java/qna/domain/DeleteHistoryTest.java:
```diff
@@ -13,19 +15,27 @@
 
 @DataJpaTest
 @EnableJpaAuditing
```
Auditing 어노테이션이 테스트마다 쓰이는 것 같은데 분리해보면 어떨까요~?

victor-jo replied:
ㅠ 잘 모르겠어서 일단 Import 로 하나의 Configuration 만 사용하도록 묶어두었습니다!

minzzang commented on src/test/java/qna/domain/DeleteHistoryTest.java:
```diff
@@ -13,19 +15,27 @@
-    public static final DeleteHistory DH1 = new DeleteHistory(ContentType.ANSWER, 1L, 1L, LocalDateTime.now());
-    public static final DeleteHistory DH2 = new DeleteHistory(ContentType.QUESTION, 2L, 1L, LocalDateTime.now());
+    public static final DeleteHistory DH1 = new DeleteHistory(ContentType.ANSWER, 1L, UserTest.JAVAJIGI, LocalDateTime.now());
+    public static final DeleteHistory DH2 = new DeleteHistory(ContentType.QUESTION, 2L, UserTest.JAVAJIGI, LocalDateTime.now());
```
test를 작성할 때 최대한 독립적으로 만들어보면 어떨까요~? UserTest.JAVAJIGI를 사용하는 것보단 DeleteHistory에서 작성하거나 따로 fixture를 분리해 보는 것도 좋을 것 같아요. static 변수는 여러 테스트에서 공유하기 때문에 독립적인 테스트를 만들기 어려울 것 같습니다! 그러면 TruncateConfig도 없앨 수 있을 것 같아요 ! 👍

victor-jo replied:
넵 좋습니다!! static 변수와 관련없게 독립테스트로 모두 만들었습니다!

minzzang reviewed (changes requested):
안녕하세요 원준님~!

2단계 미션도 잘 작성해주셨네요! 👍 

소소한 코멘트 남겼는데 확인 부탁드립니다. :pray:

그럼 오늘 하루도 화이팅입니다. :muscle:

victor-jo committed:
refactor(answer, deleteHistory): 붎필요한 메소드 제거

- 불필요한 Equals, Hashcode 제거

victor-jo committed:
refactor(answerTest): 테스트 방법 리팩터링

- static 으로 선언된 필드를 사용하지 않고 자체 테스트

victor-jo committed:
refactor(deleteHistoryTest): 테스트 방법 리팩터링

- static 으로 선언된 필드를 사용하지 않고 자체 테스트

victor-jo committed:
refactor(answerTest): 답변 엔티티 테스트 코드 리팩터링

- 중복된 코드 제거, 간결화

victor-jo committed:
refactor(questionTest): 테스트 방법 리팩터링

- static 으로 선언된 필드를 사용하지 않고 자체 테스트

victor-jo committed:
refactor(userTest): 테스트 방법 리팩터링

- static 으로 선언된 필드를 사용하지 않고 자체 테스트

victor-jo committed:
refactor(deleteHistory): 삭제 히스토리 Equals 기준 정립

- QnA 서비스 테스트코드에서 요구하는 ContentType, ContentId, Writer 를 기준으로 Equals 오버라이드
- LocalDateTime (createdAt) 의 경우 무시하도록 수정

victor-jo committed:
refactor(noNeed): 불필요한 설정 파일 제거

- static 필드 공유로 생기는 문제 해결하여 TruncateConfig 제거

victor-jo committed:
refactor(test): 테스트 코드 리팩터링

- EnableJpaAuditing 어노테이션을 계속 사용하는 것에 대해 Configuration 주입 방법으로 변경

victor-jo committed:
refactor(answers): 답변 리스트 엔티티 리팩터링

- 일급 컬렉션으로 묶어서 답변 리스트를 객체화
- 답변 리스트에 관련한 메시지 추가

victor-jo commented:
피드백 반영하여 리뷰 요청드립니다! 잘 부탁드립니다!

minzzang commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,56 @@
+import java.util.stream.Collectors;
+
+@Embeddable
+public class Answers implements Iterable<Answer> {
```
일급 컬렉션 👍 Iterable을 구현한 이유는 무엇인가요~?

minzzang reviewed (approved):
안녕하세요 원준님~!
피드백 잘 반영해주셨네요! :+1:

소소한 코멘트 남겼는데 확인 부탁드립니다. :pray:

그럼 다음 미션도 화이팅입니다. :muscle:
